Sources/DistributedCluster/Plugins/ClusterJournal/ClusterJournalPlugin.swift

```diff
@@ -0,0 +1,125 @@
+    try await store.persistEvent(event, id: persistenceId)
+  }
+  
+  /// As we already checked whenLocal on `actorReady`—would be nice to have some type level understanding already here and not to double check...
+  public func restoreEventsFor<A: EventSourced>(actor: A, id persistenceId: PersistenceID) async {
+    /// Checking if actor is already in restoring state
+    guard self.restoringActorTasks[persistenceId] == .none else { return }
+    self.restoringActorTasks[persistenceId] = Task { [weak actor, weak self] in
+      defer { Task { [weak self] in await self?.removeTaskFor(id: persistenceId) } }
+      do {
+        let events: [A.Event] = (try await self?.store.eventsFor(id: persistenceId)) ?? []
+        await actor?.whenLocal { myself in
+          for event in events {
+            myself.handleEvent(event)
+          }
+        }
+      } catch {
+        await self?.actorSystem.log.error(
+          "Cluster journal haven't been able to restore state of an actor \(persistenceId), reason: \(error)"
+        )
+      }
+    }
+  }
+  
+  private func finishContinuationsFor(
+    id: PersistenceID
+  ) {
+    for emit in (self.emitContinuations[id] ?? []) { emit.resume() }
+    self.emitContinuations.removeValue(forKey: id)
+  }
+  
+  private func removeTaskFor(id: PersistenceID) {
+    self.restoringActorTasks.removeValue(forKey: id)
+    self.finishContinuationsFor(id: id)
+  }
+    
+  public init(
+    factory: @escaping (ClusterSystem) -> any EventStore
+  ) {
+    self.factory = factory
+  }
+}
+
+extension ClusterJournalPlugin: _Plugin {
+  static let pluginKey: Key = "$clusterJournal"
+  
+  public nonisolated var key: Key {
+    Self.pluginKey
+  }
+  
+  public func start(_ system: ClusterSystem) async throws {
+    self.actorSystem = system
+    self.store = try await system
+      .singleton
+      .host(name: "\(ClusterJournalPlugin.pluginKey)_store") {
+        try await AnyEventStore(actorSystem: $0, store: self.factory($0))
+      }
+  }
+  
+  public func stop(_ system: ClusterSystem) async {
+    self.actorSystem = nil
+    self.store = nil
+    for task in self.restoringActorTasks.values { task.cancel() }
+  }
+}
+
+extension ClusterSystem {
+  
+  public var journal: ClusterJournalPlugin {
+    let key = ClusterJournalPlugin.pluginKey
+    guard let journalPlugin = self.settings.plugins[key] else {
+      fatalError("No plugin found for key: [\(key)], installed plugins: \(self.settings.plugins)")
+    }
+    return journalPlugin
+  }
+}
+
+extension EventSourced where ActorSystem == ClusterSystem {
+  public func emit(event: Event) async throws {
+    try await self.whenLocal { [weak self] myself in
+      try await self?.actorSystem.journal.emit(event, id: myself.persistenceId)
+    }
+  }
+}
+
+/// Not sure if it's correct way, basically wrapping `any EventStore` into `AnyEventStore` which is singleton
+distributed actor AnyEventStore: EventStore, ClusterSingleton {
+  
+  private var store: any EventStore
```

Comment:
Works for now, need to check and test thoroughly that it doesn't create any issues...